### ISSUE TITLE
Workflow Instance SQL Persistence 

### ISSUE THREAD

**Issue by HanaaGhasoub** [elsa-workflows/elsa-core#334]
**Scenario:** When executing an activity that completes the instance it failed because the workflow context does not have the pre-defined scope variables by the workflow activities definition while executing it through the unit test(Without SQL Persistence), it completed with "Finished" status.

**Actual:** Elsa with SQL Server does not persist in the context scopes which were defined in the workflow activities, while it is working in unit test(Without SQL Persistence).

**Expected:** Workflow complete with status "Finished".

Here the unit test context scopes:
![image](https://user-images.githubusercontent.com/30862539/85223672-e549c380-b3c4-11ea-9d44-bef872f17519.png)

And here the context scopes while doing an activity from the system itself:
![image](https://user-images.githubusercontent.com/30862539/85223833-1b3b7780-b3c6-11ea-8084-754317c7b702.png)

Please note that both were taken from the same activity source execution:
![image](https://user-images.githubusercontent.com/30862539/85223851-4a51e900-b3c6-11ea-9dad-5b7817860b56.png)

Here our workflow activities definition:
```
public class MultiStepApproval : IWorkflow, ITransientDependency{

		public void Build(IWorkflowBuilder builder)
		{
			builder
				.StartWith<WorkflowStartNotification>()
				.Then<SetVariable>(
					setup =>
					{
						setup.VariableName = "ProcessId";
						setup.ValueExpression = new JavaScriptExpression<int>("input('ProcessId')");
					})
				.Then<SetVariable>(
					setup =>
					{
						setup.VariableName = "SubmitterId";
						setup.ValueExpression = new JavaScriptExpression<long>("input('SubmitterId')");
					})
				.Then<SetVariable>(
					setup =>
					{
						setup.VariableName = "OrganizationUnitId";
						setup.ValueExpression = new JavaScriptExpression<long>("input('OrganizationUnitId')");
					})
				.Then<SetVariable>(
					setup =>
					{
						setup.VariableName = "TenantId";
						setup.ValueExpression = new JavaScriptExpression<int>("input('TenantId')");
					})
				.Then<GetWorkflowProcessActivityTypesIds>(
					setup =>
					{
						setup.WorkflowProcessId = new JavaScriptExpression<int>("ProcessId");
						setup.WorkflowProcessActivityTypeIds = "WorkflowProcessActivityTypeIds";
					})
				.Then<ForEach>(
					setup =>
					{
						setup.CollectionExpression =
							new JavaScriptExpression<IList>("WorkflowProcessActivityTypeIds");
					},
					forEach =>
					{
						forEach
							.When(OutcomeNames.Iterate)
							.Then<ResolveRelation>(setup =>
							{
								setup.ResolvedRelationResult = "ResolvedRelationResult";
								setup.AssignmentType = "AssignmentType";
								setup.UserId = new JavaScriptExpression<long>("SubmitterId");
								setup.OrganizationUnitId =
									new JavaScriptExpression<long>("OrganizationUnitId");
								setup.WorkflowActivityTypeId =
									new JavaScriptExpression<long>("CurrentValue");
							})
							.Then<SimpleApproveReject>(
								setup =>
								{
									setup.AssignmentType =
										new JavaScriptExpression<int>(
											"AssignmentType");
									setup.OrganizationUnitId =
										new JavaScriptExpression<long>("OrganizationUnitId");
									setup.WorkflowActivityTypeId =
										new JavaScriptExpression<long>("CurrentValue");
									setup.AssignToRoleId =
										new JavaScriptExpression<int>("ResolvedRelationResult");
									setup.AssignToUserId =
										new JavaScriptExpression<long>("ResolvedRelationResult");
								},
								executeHumanStep =>
								{								executeHumanStep.When(HumanApprovalOutcomeNames.Rejected)
										.Then(nameof(CompleteWorkflowInstance));									executeHumanStep.When(HumanApprovalOutcomeNames.Accepted)
										.Then("ActivityTypesLoop");
								}
							).Then(forEach);
					}).WithName("ActivityTypesLoop")
			.Then<CompleteWorkflowInstance().WithName(nameof(CompleteWorkflowInstance))
				.Then<SendNotificationForWorkflowCompletion>()
				.Then<RaiseWorkflowInstanceCompletedEvent>(
					setup =>
					{
						setup.OrganizationUnitId =
							new JavaScriptExpression<long>("OrganizationUnitId");
						setup.TenantId =
							new JavaScriptExpression<int>("TenantId");
					})
				.Then<Finish>();
		}}
 `

**Comment by sfmskywalker** [comment]
Hi @HanaaGhasoub, can you perhaps provide me with a simple console project that reproduces this issue?

**Comment by HanaaGhasoub** [comment]
Hi @sfmskywalker , sorry for the late reply.
You can find my project sample [Here](https://github.com/HanaaGhasoub/ElsaScopeTest-master) with the reported issue.
**Note** **that the scope of the global variables are being lost after getting out from ForEach activity and only the Foreach scoped variables are seen by the other activities.**
I added a comment inside the activity "**RaiseWorkflowInstanceCompletedEvent**" which should be able to access the global variable.

**Comment by iyhammad** [comment]
The problem was very interesting to me. I think the problem starts to happen from line 107 when loading the workflow to resume it. 
![image](https://user-images.githubusercontent.com/5989020/87228113-4dc20a00-c39f-11ea-8d2d-433f7c1a8dd1.png)

The scopes stack is loaded reversed and then when the workflow continues, and with the first call of WorkflowExecutionContext.EndScope(), the main scope is deleted from the stack and that explains why in last activity, we can't get the variable value.
When I reversed the stack like below, everything worked as expected and I was able to get the value at the end of the workflow.
`
Scopes = new Stack<WorkflowExecutionScope>(new Stack<WorkflowExecutionScope>(instance.Scopes));
`
